src/Diagnostics.DataProviders/KustoHeartBeatService/KustoHeartBeatService.cs

```diff
@@ -107,75 +113,97 @@ public KustoHeartBeat(string primaryCluster, string failoverCluster, KustoDataPr
             _configuration = configuration;
         }
 
-        public async Task RunHeartBeatTask(CancellationToken cancellationToken)
+        private async Task RunHeartBeatPrimary(string activityId)
         {
-            while (!cancellationToken.IsCancellationRequested) {
-                bool primaryHeartBeatSuccess = false;
-                bool failoverHeartBeatSuccess = false;
-                Exception exceptionForLog = null;
-                string activityId = Guid.NewGuid().ToString();
-                try
+            bool primaryHeartBeatSuccess = false;
+            Exception exceptionForLog = null;
+            try
+            {
+                var primaryHeartBeat = await _kustoDataProvider.ExecuteQueryForHeartbeat(_configuration.HeartBeatQuery, PrimaryCluster, _configuration.HeartBeatTimeOut, activityId);
+
+                if (primaryHeartBeat.Rows.Count >= 1)
                 {
-                    // Run kusto query with time out
-                    var primaryTask = _kustoDataProvider.ExecuteQueryForHeartbeat(_configuration.HeartBeatQuery, PrimaryCluster, _configuration.HeartBeatTimeOut, activityId);
-                    var failoverTask = _kustoDataProvider.ExecuteQueryForHeartbeat(_configuration.HeartBeatQuery, FailoverCluster, _configuration.HeartBeatTimeOut, activityId);
-                    await Task.WhenAll(new Task[] { primaryTask, failoverTask });
+                    primaryHeartBeatSuccess = true;
+                }
+            }
+            catch (Exception ex)
+            {
+                exceptionForLog = ex;
+            }
 
-                    var result = await primaryTask;
-                    if (result.Rows.Count >= 1)
-                    {
-                        primaryHeartBeatSuccess = true;
-                    }
+            if (primaryHeartBeatSuccess)
+            {
+                _ConsecutiveFailureCount = 0;
+                _ConsecutiveSuccessCount++;
+            }
+            else
+            {
+                _ConsecutiveFailureCount++;
+                _ConsecutiveSuccessCount = 0;
+            }
 
-                    var failoverResult = await failoverTask;
-                    if (failoverResult.Rows.Count >= 1)
-                    {
-                        failoverHeartBeatSuccess = true;
-                    }
+            // if not in failover state
+            //  should failover?
+            if (UsePrimaryCluster && _ConsecutiveFailureCount >= _configuration.HeartBeatConsecutiveFailureLimit)
+            {
+                UsePrimaryCluster = false;
+            } // else should stop failover
+            else if (!UsePrimaryCluster && _ConsecutiveSuccessCount >= _configuration.HeartBeatConsecutiveFailureLimit)
```

Comment:
Can we create another configuration for `ConsecutiveSuccessLimit`? For me it reduces the confusion of comparing `ConsecutiveSuccessCount` to `HeartBeatConsecutiveFailureLimit`. And also it would be nice to have tighter grain of control of success and failures. For example, if the Kusto service fails 5 consecutive times then we'll failover but Kusto has to succeed 8 consecutive times before we go back to the primary cluster. Having these knobs now will allow us to find the best configuration we need in the future as we monitor our service.